src/gluonts/dataset/arrow/dec.py

```diff
@@ -44,39 +44,36 @@ def _arrow_to_py_list_scalar(scalar: pa.ListScalar):
 @dataclass
 class ArrowDecoder:
     columns: Dict[str, int]
-    ndarray_columns: Dict[str, int]
 
     @classmethod
     def from_schema(cls, schema):
-        columns = {}
-        ndarray_columns = {}
-
-        for idx, column in enumerate(schema):
-            if column.name.endswith("._np_shape"):
-                ndarray_columns[(column.name.rsplit(".", 1)[0])] = idx
-            else:
-                columns[column.name] = idx
-
-        return cls(columns, ndarray_columns)
+        return cls(
+            [
+                column.name
+                for column in schema
+                if not column.name.endswith("._np_shape")
+            ]
+        )
 
     def decode(self, batch, row_number):
         for row in self.decode_batch(batch.slice(row_number, row_number + 1)):
             return row
 
     def decode_batch(self, batch):
-        rows = zip(*batch)
-
-        for raw_row in rows:
+        for raw_row in batch.to_pandas().to_dict("records"):
             row = {}
-            for column_name, column_idx in self.columns.items():
-                value = _arrow_to_py(raw_row[column_idx])
-
-                shape_idx = self.ndarray_columns.get(column_name)
-
-                if shape_idx is not None:
-                    shape = _arrow_to_py(raw_row[shape_idx])
-                    value = value.reshape(shape)
-
+            for column_name in self.columns:
+                value = raw_row[column_name]
+                shape = raw_row.get(f"{column_name}._np_shape")
```

Comment:
I don't think we want to do this check for every entry.

Rather, I would keep storing `ndarray_columns` and then just itereate over them.

Reply:
We would have to run some additional experiments to see whether using `ndarray_columns` really speeds things up significantly. The codebase without it seems to be more elegant. We will provide you with benchmarks as soon as we have them.